src/Routing.jl

```diff
@@ -0,0 +1,189 @@
+    # Constants
+    day_to_s = 86400
+    km²_to_m² = 1000000
+
+    output_dir = "/groups/esm/achiang/ClimaRivers.jl/data/routing/simulations/simulations_lv05/gamma_IRF"
+
+    timeseries_dir = env.dyn_env.timeseries_dir
+    dates = collect(start_date:Day(1):end_date)
+
+    routing_lv_basins_dir = env.static_env.routing_lv_basins_dir
+    attributes_df = CSV.read(
+        joinpath(env.static_env.attributes_dir, "attributes.csv"),
+        DataFrame,
+    )
+
+    a, θ = hillslope_model.shape, hillslope_model.timescale
+    t_max = hillslope_model.t_max
+
+    distribution = [
+        (t^(a - 1) * exp(-t / θ)) / (θ^a * SpecialFunctions.gamma(a)) for
+        t in 0:(t_max - 1)
+    ]
+
+    all_basin_ids = read_routing_lv_basins(
+        joinpath(routing_lv_basins_dir, "all_basin_ids.txt"),
+    )
+    # make new txt file 
+    for basin_id in all_basin_ids # routing_lv_basins
+        timeseries_df =
+            CSV.read(joinpath(timeseries_dir, "basin_$basin_id.csv"), DataFrame)
+        filtered_df =
+            filter(row -> start_date <= row[:date] <= end_date, timeseries_df)
+        basin_area =
+            attributes_df[attributes_df.HYBAS_ID .== basin_id, :area][1]
+        runoff =
+            (filtered_df[:, :sro_sum] .+ filtered_df[:, :ssro_sum]) .*
+            basin_area ./ day_to_s .* km²_to_m²
+
+        streamflow = DSP.conv(runoff, distribution)[1:size(runoff)[1], :][:]
+
+        output_df = DataFrame(date = dates, streamflow = streamflow)
+
+        CSV.write(
+            joinpath(output_dir, "hillslope_basin_$basin_id.csv"),
+            output_df,
+        )
+    end
+end
+
+
+function update_state_from_channel!(
+    river_state::RS,
+    channel_model::CM,
+    env::E,
+    start_date::Date,
+    end_date::Date,
+) where {RS <: RiverState, CM <: AbstractChannelModel, E <: Environment}
+    km_to_m = 1e3
+
+    output_dir = "/groups/esm/achiang/ClimaRivers.jl/data/routing/simulations/simulations_lv05/gamma_IRF"
+    timeseries_dir = env.dyn_env.timeseries_dir
+
+    routing_lv_basins_dir = env.static_env.routing_lv_basins_dir
+    attributes_df = CSV.read(
+        joinpath(env.static_env.attributes_dir, "attributes.csv"),
+        DataFrame,
+    )
+
+    C, D = channel_model.wave_velocity, channel_model.diffusivity
+    t_max = channel_model.t_max
+
+    all_basin_ids = read_routing_lv_basins(
+        joinpath(routing_lv_basins_dir, "all_basin_ids.txt"),
+    )
+
+    # Iterate over basins in the given routing level
+    for basin_id in all_basin_ids
+        # Get current streamflow in the basin
+        timeseries_df = CSV.read(
+            joinpath(output_dir, "hillslope_basin_$basin_id.csv"),
+            DataFrame,
+        )
+
+        # Allocate streamflow from upstreams
+        up_streamflow = zeros(length(timeseries_df[:, :streamflow]))
+
+        # Iterate over upstreams
+        for up_basin in env.graph_dict[string(basin_id)]
```

Comment:
@achiang17  Let's replace this with env.upstream_basins[string(basin_id)], where upstream_basins is a dictionary.

We think the graph_dict only has one level upstream, but it also needs all the upstream basins from that.